File: backend/src/main/java/com/nischal/backend/service/InventoryService.java
package com.nischal.backend.service;

import com.nischal.backend.dto.inventory.*;
import com.nischal.backend.entity.Order;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface InventoryService {

    // --- Ingredients ---
    List<IngredientResponse> getAllIngredients();
    List<IngredientResponse> getLowStockIngredients();
    IngredientResponse getIngredientById(Long id);
    IngredientResponse createIngredient(CreateIngredientRequest request);
    IngredientResponse updateIngredient(Long id, CreateIngredientRequest request);
    void deleteIngredient(Long id);

    // --- Stock adjustments ---
    IngredientResponse adjustStock(Long ingredientId, AdjustStockRequest request, String performedBy);

    // --- Recipe management ---
    List<RecipeIngredientResponse> getRecipeForMenuItem(Long menuItemId);
    RecipeIngredientResponse addRecipeIngredient(Long menuItemId, RecipeIngredientRequest request);
    void removeRecipeIngredient(Long menuItemId, Long ingredientId);

    // --- Auto-deduction on order ---
    void deductForOrder(Order order);

    // --- Logs ---
    Page<InventoryLogResponse> getRecentLogs(Pageable pageable);
    List<InventoryLogResponse> getLogsForIngredient(Long ingredientId);
}

File: frontend/src/services/inventory.service.ts
import { apiClient } from './api.service';
import type {
  Ingredient,
  CreateIngredientRequest,
  AdjustStockRequest,
  RecipeIngredient,
  RecipeIngredientRequest,
  InventoryLog,
} from '../types/inventory.types';

export const inventoryService = {
  async getAllIngredients(): Promise<Ingredient[]> {
    return apiClient.get<Ingredient[]>('/inventory');
  },

  async getLowStockIngredients(): Promise<Ingredient[]> {
    return apiClient.get<Ingredient[]>('/inventory/low-stock');
  },

  async createIngredient(data: CreateIngredientRequest): Promise<Ingredient> {
    return apiClient.post<Ingredient>('/inventory', data);
  },

  async updateIngredient(id: number, data: CreateIngredientRequest): Promise<Ingredient> {
    return apiClient.put<Ingredient>(`/inventory/${id}`, data);
  },

  async adjustStock(id: number, data: AdjustStockRequest): Promise<Ingredient> {
    return apiClient.post<Ingredient>(`/inventory/${id}/adjust`, data);
  },

  async deleteIngredient(id: number): Promise<void> {
    return apiClient.delete<void>(`/inventory/${id}`);
  },

  async getRecipe(menuItemId: number): Promise<RecipeIngredient[]> {
    return apiClient.get<RecipeIngredient[]>(`/inventory/recipe/${menuItemId}`);
  },

  async addRecipeIngredient(menuItemId: number, data: RecipeIngredientRequest): Promise<RecipeIngredient> {
    return apiClient.post<RecipeIngredient>(`/inventory/recipe/${menuItemId}`, data);
  },

  async removeRecipeIngredient(menuItemId: number, ingredientId: number): Promise<void> {
    return apiClient.delete<void>(`/inventory/recipe/${menuItemId}/ingredient/${ingredientId}`);
  },

  async getLogs(page = 0, size = 20): Promise<{ content: InventoryLog[]; totalElements: number }> {
    return apiClient.get(`/inventory/logs?page=${page}&size=${size}`);
  },

  async getIngredientLogs(id: number, page = 0, size = 20): Promise<{ content: InventoryLog[]; totalElements: number }> {
    return apiClient.get(`/inventory/${id}/logs?page=${page}&size=${size}`);
  },
};
